### ISSUE TITLE
ember-simple-auth integration

### ISSUE THREAD

**Issue by oliverbarnes** [pixelhandler/ember-jsonapi-resources#22]
I noticed my app wasn't sending the authorization header when creating a resource, and after some digging I got it going by overriding `ApplicationAdapter#fetchOptions` in the resource adapter, making it lookup the access token in ember-simple-auth's session:

```
fetchOptions(options) {
    let isUpdate;
    options.headers = options.headers || { 'Content-Type': 'application/vnd.api+json' };
    // const authHeader = window.localStorage.getItem('AuthorizationHeader');
    const simpleAuthSession = JSON.parse(window.localStorage.getItem('ember_simple_auth:session'));
    const accessToken       = simpleAuthSession.secure.access_token;
    const authHeader        = 'Bearer ' + accessToken;

    if (authHeader) {
      options.headers['Authorization'] = authHeader;
    }
    if (typeof options.update === 'boolean') {
      isUpdate = options.update;
      delete options.update;
    }
    return isUpdate;
  }
```

I'm sure there must be a better way :) Beginning to get familiar with ember-jsonapi-resources, and I'm really a rubyist still grokking ember. I'll run this by ember-simple-auth's author tomorrow (I work with him), to see if he has any suggestions.

Thought it worth bringing up, as people using EJR might be using ember-simple-auth as well, and this could get the ball rolling for a proper solution.

**Comment by pixelhandler** [comment]
@oliverbarnes I'll take a look at creating a method to `getAuthorizationToken` and call that from the `fetchOptions` method. Also I'll add this example to the wiki. Thanks for sharing your solution!

**Comment by oliverbarnes** [comment]
You’re welcome, I’d submit a patch if I could. In the future when I’m better versed I’ll do that.

the getAuthorizationToken approach sounds good, specially if configurable through environment.js

Oliver

On July 16, 2015 at 6:49:20 AM, Bill Heaton (notifications@github.com) wrote:

@oliverbarnes I'll take a look at creating a method to getAuthorizationToken and call that from the fetchOptions method. Also I'll add this example to the wiki. Thanks for sharing your solution!

**Comment by pixelhandler** [comment]
@oliverbarnes I'm thinking a mixin for the authorization like so:

```javascript
import Ember from 'ember';

export default Ember.Mixin.create({

  /**
    The name of the Authorization request-header field

    @property authorizationHeaderField
    @type String
    @required
  */
  authorizationHeaderField: 'Authorization',

  /**
    The name key, stored locally, that references the Authorization request-header credential/token

    @property authorizationHeaderStorageKey
    @type String
    @required
  */
  authorizationHeaderStorageKey: 'AuthorizationHeader',

  /**
    Authentication credentials/token used with HTTP authentication

    @property authorizationCredential
    @type String
    @required
  */
  authorizationCredential: Ember.computed({
    get(key) {
      key = this.get('authorizationHeaderStorageKey');
      return window[this._storage].getItem(key);
    },
    set(key, value) {
      key = this.get('authorizationHeaderStorageKey');
      window[this._storage].setItem(key, value);
    }
  }),

  /**
    Storage type localStorage or sessionStorage

    @property _storage
    @type String
    @private
  */
  _storage: ['localStorage', 'sessionStorage'][0]

});
```

Then the mixin can be reopened and the `authorizationHeaderStorageKey` set to `'ember_simple_auth:session'` and/or `authorizationCredential` `get` method set to extract the `access_token` from storage and concatenate whatever is necessary to form the token that will be assigned to the "Authorization" request-header as the credentials.

**Comment by pixelhandler** [comment]
@oliverbarnes see this solution doc and let me know if the new mixin works out for you, https://github.com/pixelhandler/ember-jsonapi-resources/wiki/Customizing-your-Authorization-Header

**Comment by oliverbarnes** [comment]
awesome, thanks @pixelhandler. I'll try it out this weekend, saw the pull request but didn't have time to look at it so far

**Comment by oliverbarnes** [comment]
Hey, finally got around to testing it, and it works for me. Just took me a couple of hoops to getting working based on docs, I'll try updating them to clarify for the next noob ;)

**Comment by oliverbarnes** [comment]
Looks like I can't edit the wiki, you might have to add me as a collaborator? I would just increment authorization mixin example to include the path where it needs to be saved, add an ember import, and close the computed property argument block:

```
//app/mixins/authorization.js

import Ember from 'ember';
import AuthorizationMixin from 'ember-jsonapi-resources/mixins/authorization';

export default AuthorizationMixin.reopen({
  authorizationHeaderStorageKey: 'ember_simple_auth:session',
  authorizationCredential: Ember.computed({
    get(key) {
      key = this.get('authorizationHeaderStorageKey');
      const simpleAuthSession = JSON.parse(window.localStorage.getItem(key));
      return 'Bearer ' + simpleAuthSession.secure.access_token;
    }
  })
});
```

I'd then add an example of where it's being used by an adapter:

```
import ApplicationAdapter from 'ember-jsonapi-resources/adapters/application';
import AuthorizationMixin from '../mixins/authorization';
import config from '../config/environment';

export default ApplicationAdapter.extend(AuthorizationMixin, {
  type: 'proposal',
  url: config.APP.API_HOST +  'proposals'
});
```

BTW, is it possible for me to re-open ApplicationAdapter to include the mixin internally, so all adapters have it by default?

**Comment by pixelhandler** [comment]
@oliverbarnes I added you as a collaborator, feel free to update the wiki pages, and thank you for testing out the change for auth support :)

Regarding all adapters using the mixin, all I had to do what change the path for the import, I had to also change the generator blueprint too, see this example https://github.com/pixelhandler/jr-test/commit/8694c5a9dfbe179de92388f890050774988d91ed

Loading the application adapter from the relative path at `./application` uses the application adapter with the mixin applied to it. Previously I imported the application adapter from the addon directory not the app directory. So for all adapters that extend the application adapter the auth mixin is already applied.
